accepted/2021/experimental-features/experimental-features.md

```diff
@@ -0,0 +1,390 @@
+[assembly: ExperimentalFeature("FancyGC")]
+```
+
+This means you can look at a binary and tell which experimental features were
+used to build it. This can be used for auditing and for detecting transitive
+dependencies ([discussed later](#api-analyzer)).
+
+### runtime.json
+
+The `ExperimentalFeature` should be written into the `runtime.json` file such
+that the host/runtime can use it to make decisions.
+
+### Compilation context
+
+The `ExperimentalFeature` needs to be passed to the compilers. Roslyn (C#/VB)
+already has a mechanism for this which is exposed to both the compiler as well
+as for analyzers/source generators. Alternatively, we could say the information
+about experimental features can be retrieved from the generated assembly-level
+attributes.
+
+***Note:*** F# doesn't have this capability today.
+
+### API Analyzer
+
+We'll add an attribute that we can put on APIs we consider experimental. In some
+cases it might not be possible to mark an API (for example, because it's a new
+type system feature) in which case the feature would either be guarded by new
+language syntax or by a custom analyzer.
+
+```C#
+namespace System.Runtime.Versioning
+{
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Module |
+                    AttributeTargets.Class |
+                    AttributeTargets.Struct |
+                    AttributeTargets.Enum |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Method |
+                    AttributeTargets.Property |
+                    AttributeTargets.Field |
+                    AttributeTargets.Event, AllowMultiple = true, Inherited = false)]
+    public sealed class ExperimentalFeatureAttribute : Attribute
+    {
+        public ExperimentalFeatureAttribute(string featureName);
+        public string FeatureName { get; }
+    }
+}
+```
+
+We'll ship a built-in analyzer that will report diagnostics when an API is being
+called that is marked with this attribute and the feature name is not contained
+in the project's `ExperimentalFeature` items. The default severity is `error`.
+
+This analyzer will also validate that all referenced assemblies only have
+assembly-level attributes for experimental features that the consuming project
+also declared. For example, if a project has this configuration:
+
+```xml
+<Project>
+
+  <ItemGroup>
+    <EnableExperimentalFeature Include="StaticInterfaceMembers" />
+  </ItemGroup>
+
+</Project>
+```
+
+consuming an assembly with these attributes
+
+```C#
+[assembly: ExperimentalFeature("StaticInterfaceMembers")]
+[assembly: ExperimentalFeature("FancyGC")]
+```
+
+would result in a diagnostic like this:
+
+> error: Assembly 'Foo.Bar' depends on the experimental feature 'FancyGC' which
+> the consuming project has not enabled.
+
+***Note:*** Since F# doesn't support analyzers, we would have to add this
+capability to the compiler itself.
+
+## Q & A
```

Comment:
Should the Intellisense and the debugger show all types/APIs, or filter based on the current preview features that are enabled?
To increase adoption of preview features, we may not want to filter the APIs. But we may want to display them differently, to avoid annoyance (I keep taking a dependency that I'm not allowed to).